### PR TITLE
Change the behavior of persisted index for snapshot

### PR DESCRIPTION
**Pull request by gengliqi**
Signed-off-by: gengliqi <gengliqiii@gmail.com>

In previous async ready PR(https://github.com/tikv/raft-rs/pull/403), After getting a snapshot and calling the `restore` function, the `persisted` is changed to the index of this snapshot. This makes the `persisted` index does not mean the data before this index is _really_ persisted, which is not intuitive.
The reason for this behavior is to maintain a invariant that `applied <= min(committed, persisted)`. However, I find it's not needed if the application calls the `advance_append` or `on_persist_ready` then calls `advance_apply_to`. This is intuitive because the snapshot is persisted does not mean it is applied so it should be persisted first, then be applied.

### CODE REVIEW

**Comment by gengliqi** [comment]
@BusyJay @NingLin-P PTAL, thanks.

**Comment by BusyJay** [comment]
So if the `persisted` is 2048 and snapshot reset first index back to 2040, then following new committed and not saved entries will be fetched to ready directly?

**Comment by gengliqi** [comment]
> So if the `persisted` is 2048 and snapshot resets last index back to 2040, then following new committed and not saved entries will be fetched to ready directly?

It won't happen because the `persisted` should be changed when appending new entries.

It seems it's hard to make the `persisted` to be *really persisted index*. Maybe it's no need to change?

**Comment by BusyJay** [comment]
> It won't happen because the persisted should be changed when appending new entries.

It's not true. Entries after snapshot won't cause conflict, so `persisted` won't be changed.

**Comment by gengliqi** [comment]
> > It won't happen because the persisted should be changed when appending new entries.
> 
> It's not true. Entries after snapshot won't cause conflict, so `persisted` won't be changed.

If there is no conflict, the persisted is also be changed.
https://github.com/tikv/raft-rs/blob/fc1ef2f2d9711f7b9e88fe4399445b135d5f8023/src/raft_log.rs#L227-L232

**Comment by BusyJay** [comment]
If there is no conflict, `conflict_idx` will be zero, so the else branch will be skipped.

https://github.com/tikv/raft-rs/blob/fc1ef2f2d9711f7b9e88fe4399445b135d5f8023/src/raft_log.rs#L217-L236

**Comment by gengliqi** [comment]
> If there is no conflict, `conflict_idx` will be zero, so the else branch will be skipped.
> 
> https://github.com/tikv/raft-rs/blob/fc1ef2f2d9711f7b9e88fe4399445b135d5f8023/src/raft_log.rs#L217-L236

Hmmm, if it's zero, there is no new entries.

**Comment by gengliqi** [comment]
By the way, I think we should set `persisted` to make the logic more intuitive, but set it to `index - 1` seems not good enough, I have no better idea.

**Comment by BusyJay** [comment]
How about setting it to zero, which means unknown? Or use `Option<NonZeroU64>` and make it `None`.

**Comment by gengliqi** [comment]
> How about setting it to zero, which means unknown? Or use `Option<NonZeroU64>` and make it `None`.

It has problems because there is an invariant that `applied <= min(committed, persisted)`.

After thinking for a while, I believe the most suitable method is `if self.persisted > self.committed { self.persisted = self.committed; }` so we can say the `persisted` is the **really** persisted index all the time. You can see the comments in code for details.

Also, I add test for the case mentioned before, which is 
1. Append some entries(index 20) and persist them
2. Receive a snapshot(index 10) then some new entries(index to 13)
3. Get a ready then check it(if there is a snapshot and some new entries can be applied, the code should panic in `RawNode::ready` 
```
assert!(
    !raft
        .raft_log
        .has_next_entries_since(self.commit_since_index),
     "has snapshot but also has committed entries since {}",
     self.commit_since_index
);
```

**Comment by gengliqi** [comment]
@BusyJay @NingLin-P PTAL again, thanks.

File: src/raft.rs [hunk 1]
```diff
@@ -971,10 +971,15 @@ impl<T: Storage> Raft<T> {
...
+    /// Notifies that these raft logs or snapshot have been persisted.
     pub fn on_persist_entries(&mut self, index: u64, term: u64) {
         let update = self.raft_log.maybe_persist(index, term);
-        if update && self.state == StateRole::Leader {
+        // Actually, if it is a leader and persisted index is updated, this term
+        // must be equal to self.term because the persisted index must be equal to
+        // the last index of entries from previous leader when it becomes leader
+        // (see the comments in become_leader), namely, the new persisted entries
+        // must come from this leader. Here checking the term just for robustness.
+        if update && self.state == StateRole::Leader && term == self.term {
```
Comment (BusyJay) [inline src/raft.rs:982]: How about using a log? Even if the term doesn't match `self.term` in the future adaption, for example introducing paging, it's still safe to enter the if branch.

File: src/raw_node.rs [hunk 2]
```diff
@@ -565,6 +569,11 @@ impl<T: Storage> RawNode<T> {
             }
             let mut record = self.records.pop_front().unwrap();
 
+            if let Some((i, t)) = record.snapshot {
+                index = i;
```
Comment (BusyJay) [inline src/raw_node.rs:573]: Any case to cover?

**Review by BusyJay — COMMENTED** [review]
Rest LGTM

**Review by NingLin-P — DISMISSED** [review]
LGTM

File: src/raft.rs (hunk 1 above)
Comment (gengliqi) [reply src/raft.rs]: Done.

File: src/raw_node.rs (hunk 2 above)
Comment (gengliqi) [reply src/raw_node.rs:573]: Added by changing `test_async_ready_follower`

**Review by BusyJay — APPROVED** [review]

**Review by NingLin-P — APPROVED** [review]

**Pull request merged**